Use udev rule instead of devlinks code

jbaublitz opened the pull request:
Closes https://github.com/stratis-storage/project/issues/95.
Closes #1987.

Supersedes #2098

jbaublitz commented:
@mulkieran This is still not quite ready for review, but I've made significant progress.

1. I found a way to log to journald directly from the executable so error reporting should show up there now.
2. I squashed the bug we discussed in stand up using a `filter_map` approach.
3. The udev rule is now up to date with the new expected path.
4. I've determined that udev must track which symlinks are associated with which devices because I removed the udev rule before destroying the Stratis filesystem I was testing on and udev still knew to clean up the symlinks.

Overall from a testing perspective, this seems to be working but I'm going to put this on hold for now to work with libcryptsetup-rs.

jbaublitz commented:
TODO: Resolve [this comment](https://github.com/stratis-storage/stratisd/pull/2091#discussion_r443702615).

jbaublitz commented:
Rebased.

jbaublitz commented:
I found a way to remove the `systemd` crate dependency which is causing the build to fail. This would pull in `libsystemd` which I really don't think we need to require. Working on updates to use `/dev/log` directly instead.

jbaublitz commented:
Rebased.

jbaublitz commented:
Rebased.

jbaublitz commented:
Rebased.

mulkieran commented on src/engine/strat_engine/devlinks.rs:
```diff
@@ -3,169 +3,97 @@
-
-        for leftover in existing_dirs {
-            pool_removed(&Name::new(leftover));
+pub fn pool_renamed(pool_name: &str, _: &str) {
```
At this point, the second parameter _can_ be dropped altogether.

jbaublitz replied:
Resolved.

mulkieran commented on src/engine/strat_engine/devlinks.rs:
```diff
@@ -3,169 +3,97 @@
-/// directory.
-pub fn filesystem_added(pool_name: &str, fs_name: &str, devnode: &Path) {
-    let p = filesystem_mount_path(pool_name, fs_name);
+pub fn filesystem_renamed(pool_name: &str, fs_name: &str, _: &str) {
```
Must as well drop this unused parameter, also.

jbaublitz replied:
Resolved.

mulkieran commented on src/engine/strat_engine/devlinks.rs:
```diff
@@ -3,169 +3,97 @@
-/// directory.
-pub fn filesystem_added(pool_name: &str, fs_name: &str, devnode: &Path) {
-    let p = filesystem_mount_path(pool_name, fs_name);
+pub fn filesystem_renamed(pool_name: &str, fs_name: &str, _: &str) {
```
Please put in a header comment, and also for method above.

jbaublitz replied:
Resolved.

jbaublitz commented on Cargo.toml:
```diff
@@ -28,6 +33,10 @@ timerfd = "1.0.0"
 itertools = "0.8.0"
 libcryptsetup-rs = "0.4.0"
 
+[dependencies.regex]
```
Resolved.

mulkieran replied:
> Just so you're aware, this won't change the behavior at all, but I agree it's more readable. Fixing.

Yeah. It's just programmer behavior that I'm trying to affect. Anything that makes it a clearer read.

jbaublitz commented:
Rebased.

jbaublitz commented:
@mulkieran Take a special look at the most recent commit as it's removing cruft that is no longer needed in the engine API with the udev script.

mulkieran commented:
> @mulkieran Take a special look at the most recent commit as it's removing cruft that is no longer needed in the engine API with the udev script.

Great!

mulkieran commented:
@jbaublitz Please rebase, thanks.

jbaublitz commented:
@mulkieran Rebased.

jbaublitz commented:
@mulkieran Ok, that should be resolved!

mulkieran commented on src/bin/stratis_uuids_to_names.rs:
```diff
@@ -0,0 +1,201 @@
+    };
+    match args.next().as_deref() {
+        Some(action) => {
+            if action == "remove" {
```
I get this from udevadm:

```
$ udevadm test --action=help
add
remove
change
move
online
offline
bind
unbind
```

I _think_ we'll probably want to exclude them all except for ```add``` and ```change```.

mulkieran commented on src/bin/stratis_uuids_to_names.rs:
```diff
@@ -0,0 +1,201 @@
+use uuid::Uuid;
+
+pub const DEV_LOG: &str = "/dev/log";
+pub const SYSTEM_DAEMON_INFO: &str = "<30>";
```
I couldn't figure out where these two numbers come from or why they're in brackets.

mulkieran commented on src/bin/stratis_uuids_to_names.rs:
```diff
@@ -0,0 +1,201 @@
+            .ok_or_else(|| StratisUdevError::new("Could not get pool name from UUID."))?;
+        let fs_name = uuid_to_stratis_name(&managed_objects, STRATIS_FS_IFACE, fs_uuid)?
+            .ok_or_else(|| StratisUdevError::new("Could not get filesystem name from UUID."))?;
+        println!("{} {}", pool_name, fs_name);
```
Can you swap the println! into the ```main``` method? It seems a bit tidier.

mulkieran commented on src/bin/stratis_uuids_to_names.rs:
```diff
@@ -0,0 +1,201 @@
+        Ok(Some((pool_name, fs_name))) => {
+            sock.send_to(
+                format!(
+                    "{}Symlink /dev/stratis/{}/{} succesfully created.",
```
Can you drop the "succesfully"? "created" should be enough.

jbaublitz commented:
@mulkieran Should be ready for another pass.

mulkieran commented on src/bin/stratis_uuids_to_names.rs:
```diff
@@ -0,0 +1,206 @@
+    let mut names: Vec<_> = managed_objects
+        .values()
+        .filter_map(|map| {
+            if map.contains_key(iface_name)
```
Should be able to drop ```contains_key``` here, since using ```get```.

mulkieran commented on src/bin/stratis_uuids_to_names.rs:
```diff
@@ -0,0 +1,206 @@
+    managed_objects: &GMORet,
+    iface_name: &'static str,
+    uuid: Uuid,
+) -> Result<Option<String>, StratisUdevError> {
```
I think ```Option``` part of return type is unnecessary; either there is an error or the name is found.

jbaublitz commented on src/bin/stratis_uuids_to_names.rs:
```diff
@@ -31,7 +31,11 @@ use regex::Regex;
 
 pub const DEV_LOG: &str = "/dev/log";
+/// Syslog priority syntax
+/// (3 (SYSTEM) << 3) | 6 (INFO)
```
That is part of the syntax. It is required.

jbaublitz replied:
(Just to clarify, if the brackets are not provided, the number will be added to the log message instead of being interpreted as a priority)

mulkieran replied:
ok, thx!

jbaublitz commented on udev/99-stratisd.rules:
```diff
@@ -0,0 +1 @@
+ENV{DM_NAME}=="stratis-1-[0-9a-f]*-thin-fs-[0-9a-f]*", PROGRAM="/usr/lib/udev/stratis_uuids_to_names $env{DM_NAME} $env{ACTION}", SYMLINK+="stratis/%c{1}/%c{2}"
```
What is the concern here? I'm unclear on what potential problems you see this causing. And what kind of indirection would you see being helpful? A symlink? Use of a variable provided by udev internally? I think that because we are providing that executable just like we are providing the udev rule, there is no ambiguity in the absolute path that we're providing so if the package is installed, one should be able to access the other without a problem.

bgurney-rh replied:
The macro `%{_libdir}` evaluates to `/usr/lib64` on an x86-64 system, yet the correct path to use for this file is `/usr/lib/udev`.  This may be only specific to the spec file, for packaging, but it does call into question why there's a `%{_udevrulesdir}` macro (which resolves to `/usr/lib/udev/rules.d`), and not one for `/usr/lib/udev`.

mulkieran replied:
@bgurney-rh has hit on the area of my concern; it's unforeseen interactions between Fedora/RHEL packaging and our hard-coding. Right now there are none; because the spec file also hard codes the path: https://github.com/stratis-storage/ci/pull/106/commits/b82400732e14201f10bf28b30f0cb32b9086a123. So we can proceed.

jbaublitz replied:
I'm going to test removing the absolute path and using a relative one instead, and I'll report back.

jbaublitz replied:
This works.

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -438,7 +430,6 @@ mod test {
         assert_eq!(engine.liminal_devices, LiminalDevices::default());
 
         engine.teardown().unwrap();
-        remove_dir_all(DEV_PATH).unwrap();
```
Could you edit the test comment above as well? Where it says remove "/stratis".

jbaublitz commented:
Rebased and PR revisions made. @mulkieran I changed on of the commits slightly to keep the accesses of `submap` all at the same closure level just to make it a little bit more readable but other than that. If that works for you, this should be ready to go.
